File: app/src/main/java/com/omar/retromp3recorder/app/di/UtilsModule.java
package com.omar.retromp3recorder.app.di;

import android.content.Context;

import com.omar.retromp3recorder.app.stringer.ContextStringer;
import com.omar.retromp3recorder.app.stringer.Stringer;

import javax.inject.Named;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Scheduler;

import static com.omar.retromp3recorder.app.di.AppComponent.MAIN_THREAD;
import static io.reactivex.android.schedulers.AndroidSchedulers.mainThread;
import static io.reactivex.schedulers.Schedulers.io;

@Module
public class UtilsModule {

    private final Context context;

    public UtilsModule(Context context) {
        this.context = context;
    }

    @Provides
    Scheduler provideScheduler() {
        return io();
    }

    @Provides
    @Named(MAIN_THREAD)
    Scheduler provideMainThreadScheduler(){
        return mainThread();
    }

    @Provides
    Context context() {
        return context;
    }

    @Provides
    Stringer providerStringer(Context context) {
        return new ContextStringer(context);
    }
}

File: app/src/main/java/com/omar/retromp3recorder/app/di/UtilsModule.kt
package com.omar.retromp3recorder.app.di

import android.app.Application
import android.content.Context
import com.omar.retromp3recorder.app.WakelockDealerImpl
import com.omar.retromp3recorder.utils.*
import dagger.Module
import dagger.Provides
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Named

@Module
class UtilsModule(private val app: Application) {
    @Provides
    fun provideScheduler(): Scheduler {
        return Schedulers.io()
    }

    @Provides
    @Named(MAIN_THREAD)
    fun provideMainThreadScheduler(): Scheduler {
        return AndroidSchedulers.mainThread()
    }

    @Provides
    fun context(): Context {
        return app
    }

    @Provides
    fun provideFilePathGenerator(filePathGeneratorImpl: FilePathGeneratorImpl): FilePathGenerator =
        filePathGeneratorImpl

    @Provides
    fun provideFileLister(fileListerImpl: FileListerImpl): FileLister = fileListerImpl

    @Provides
    fun provideFileNonEmptyChecker(fileEmptyCheckerImpl: FileEmptyCheckerImpl): FileEmptyChecker =
        fileEmptyCheckerImpl

    @Provides
    fun providePermissionChecker(permissionCheckerImpl: PermissionCheckerImpl): PermissionChecker =
        permissionCheckerImpl

    @Provides
    fun provideFileDeleter(fileDeleter: FileDeleterImpl): FileDeleter = fileDeleter

    @Provides
    fun provideFileRenamer(fileRenamer: FileRenameImpl): FileRenamer = fileRenamer

    @Provides
    fun provideDirCreator(dirCreatorImpl: DirCreatorImpl): DirCreator = dirCreatorImpl

    @Provides
    fun provideMp3MetadataEditor(mp3TagsEditorImpl: Mp3TagsEditorImpl): Mp3TagsEditor =
        mp3TagsEditorImpl

    @Provides
    fun provideWakelockDealer(wakelockDealerImpl: WakelockDealerImpl): WakeLockDealer =
        wakelockDealerImpl
}
